New firewall functions

ernierasta opened the pull request:
Hi!

So I have something. It should be quite finished. I need to test it more and get rid of some TODO's. Any comments, ideas - very welcome. Unfortunately there is no easy way to show You godoc for branch, but is almost finished, I need to polish it and verify (and add examples for `FirewallAdvancedRuleAdd`).

Mentioned `FirewallAdvancedRuleAdd` is very powerful, but dangerous one, it is also accepting FWRule struct as parameter (which You can create or get by FirewallGetRules). Rest of functs accept simple strings and numbers.

If You miss any function let me know, maybe we should have also `FirewallGetRule(name string)(FWRule, error)`?

Sorry for unknown as author, I am usually not writing code on Windows and my git was not configured.

ernierasta commented:
Hi, I think now it is ok. Sorry for so big PR, but I do not think it would make sense to split it in this case. Any comments welcome.

iamacarpet commented on firewall.go:
```diff
@@ -139,3 +667,67 @@ func FirewallRuleExistsByName(rules *ole.IDispatch, name string) (bool, error) {
+// apiInit initialize common fw api.
+// then:
+// dispatch apiRelease(u, fwp)
+func apiInit() (*ole.IUnknown, *ole.IDispatch, error) {
```
We are in a global namespace here with the rest of the library, so do you think this function name should include "firewall", to help make sure it doesn't get confusing?

ernierasta commented:
Hi,
definitely firewall prefix should be there!
And yes, FirewallRuleGet would be better. I had renamed FirewallGetRules to FirewallRulesGet for that reason.

I usually develop in Linux with vim, so I have everything set there, if my VS Code did not run gofmt i will run it at evening and push mentioned changes and FirewallRuleGet.

ernierasta commented:
Btw I just run:
`gofmt -d firewall.go` and looks like it gofmt-ed (no output). To my eye it also looks good, do You have something specific in mind which looks bad?

iamacarpet commented on firewall.go:
```diff
@@ -139,3 +667,67 @@ func FirewallRuleExistsByName(rules *ole.IDispatch, name string) (bool, error) {
+}
+
+// apiRelease cleans memory.
+func apiRelease(u *ole.IUnknown, fwp *ole.IDispatch) {
```
same as above, probably best to add some "firewall" naming here since it is in the global namespace.

ernierasta replied:
done

iamacarpet commented on firewall.go:
```diff
@@ -10,33 +10,520 @@ import (
+//   NET_FW_PROFILE2_CURRENT // adds rule to currently used FW Profile
+//   NET_FW_PROFILE2_ALL // adds rule to all profiles
+//   NET_FW_PROFILE2_DOMAIN|NET_FW_PROFILE2_PRIVATE // rule in Private and Domain profile
+func FirewallApplicationRuleAdd(name, description, group, appPath string, profile int32) (bool, error) {
```
`FirewallRuleAddApplication` ?

ernierasta replied:
done

iamacarpet commented on firewall.go:
```diff
@@ -10,33 +10,520 @@ import (
+// in given profile. So use with caution.
+//
+// HINT: Use FirewallRulesGet to get examles how rules can be defined.
+func FirewallAdvancedRuleAdd(rule FWRule) (bool, error) {
```
`FirewallRuleAddAdvanced` ?

ernierasta replied:
done

ernierasta commented:
Sorry for no reaction, i am starting my vacations today so work catch me up. I had no time to look at, but I agree with all mentioned changes. I was thinking about f.e: FirewallRuleAddAdvanced, maybe even better:
`FirewallRuleAdvancedAdd`
so we have:
TopicObjectTypeAction

I will be away for 3 weeks (we will bike north part of Czech Republic :-) ), so if You want you can merge it as is and make changes or wait until I will be back.

Great weekend to You too!

iamacarpet commented:
Sounds like a pretty awesome vacation, very jealous :).

I'm in no rush so happy to wait until you are back.

Hope you have a great time!

ernierasta commented:
Hi! I am back. My vacations was not so long, but work, other things ... You know :-) . I had added added mentioned function and fixed typo and 2 private function names. I will push rest of function names changes today later.

BTW: Vacations was awesome! Weather was good and we get used to pretty steep hills quickly. And beer was great as always. ;-)

ernierasta commented:
Ok, done. Ready to merge? ;-)

iamacarpet commented:
Hello again @ernierasta ,

Great to hear the holiday was nice! What I'd give for some country side and a beer, you are living the dream! :).

All looks good to me, just noticed the file name "firewall_examles_test.go" has a typo in it, but I can always fix that once it's merged if you'd like?

Thanks again for all your efforts.

Regards,
iamacarpet

ernierasta commented:
Hehe, yeah, can not complain. :-)

Fixed type in file name. I am afraid, there may other typos here and there ... well that is me. :-)
